src/util/text-tracks.js

```diff
@@ -223,6 +223,59 @@ export const addMetadata = ({
   });
 };
 
+/**
+ * Add daterange metadata text track to a source handler given an array of metadata
+ *
+ * @param {Object}
+ *   @param {Object} inbandTextTracks the inband text tracks
+ *   @param {Array} metadataArray an array of meta data
+ * @private
+ */
+export const addDaterangeMetadata = ({
+  inbandTextTracks,
+  metadataArray,
+  timestampOffset
+}) => {
+  if (!metadataArray) {
+    return;
+  }
+
+  const Cue = window.WebKitDataCue || window.VTTCue;
+  const metadataTrack = inbandTextTracks.metadataTrack_;
+  const daterangeAttr = {
+    id: 'ID',
+    class: 'CLASS',
+    startDate: 'START-DATE',
+    duration: 'DURATION',
+    endDate: 'END-DATE',
+    endOnNext: 'END-ON-NEXT',
+    plannedDuration: 'PLANNED-DURATION',
+    scte35Out: 'SCTE35-OUT',
+    scte35In: 'SCTE35-IN'
+  };
+
+  if (!metadataTrack) {
+    return;
+  }
+
+  metadataArray.daterange.forEach((metadata) => {
+    const startTime = timestampOffset + (new Date(metadata.startDate).getTime() - new Date(metadataArray.dateTimeObject).getTime()) / 1000;
+    const endTime = startTime + parseFloat(metadata.plannedDuration || metadata.duration) || 0;
```

Comment:
I think we may want to apply the `dateTimeObject` offset to the end time as well and maybe not use the `plannedDuration` for the endTime calculation. HLS.js is doing this: https://github.com/video-dev/hls.js/blob/78b50130bbacb268f3400a8211f6fe9580afb0fa/src/controller/id3-track-controller.ts#L322

Reply:
Updated endTime value in the latest commit. 
What should the endTime value be when there's no endDate attribute?

Reply:
Good question, I think 0 is probably fine, but I think we will also want to account for `END-ON-NEXT` tags.

Reply:
Updated endTime calculation and added tests in the latest commit